libraries/AP_ESC_Telem/AP_ESC_Telem_HobbyWing.cpp

```diff
@@ -0,0 +1,165 @@
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("X4_MASK",  22, AP_ESC_Telem_HobbyWing, channel_mask_x4, 0),
+
+    // @Param: X4_POLES
+    // @DisplayName: HobbyWingESC XRotor v4 Motor Poles
+    // @Description: This allows calculation of true RPM from ESC's eRPM. The default is 14.
+    // @Range: 1 127
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("X4_POLES",  23, AP_ESC_Telem_HobbyWing, motor_poles_x4, 14),
+#endif  // AP_HOBBYWING_XROTOR_V4_ENABLED
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_ESC_Telem_HobbyWing::AP_ESC_Telem_HobbyWing()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AP_ESC_Telem_HobbyWing::thread_main()
+{
+    // initialise all escs; this will set the UART up in the thread
+    for (uint8_t i=0; i<num_escs; i++) {
+        escs[i]->init();
+    }
+
+    while (true) {
+        hal.scheduler->delay_microseconds(1500);
+        for (uint8_t i=0; i<num_escs; i++) {
+            escs[i]->update();
+        }
+    }
+}
+
+void AP_ESC_Telem_HobbyWing::init()
+{
+    for (uint8_t i=0; i<32 && num_escs<ARRAY_SIZE(escs); i++) {
+        // work out our servo channel offset.  First instance of this
+        // class gets the channel corresponding to the first bit set in
+        // the mask etc
+        uint8_t version = 0;
+        uint8_t poles;
+        if (((uint32_t)channel_mask_v3 & (1U<<i))) {
+            version = 3;
+            poles = motor_poles_v3;
+        } else if (((uint32_t)channel_mask_v4 & (1U<<i))) {
+            version = 4;
+            poles = motor_poles_v4;
+        } else if (((uint32_t)channel_mask_x4 & (1U<<i))) {
+            version = 5;
+            poles = motor_poles_x4;
+        } else {
+            continue;
+        }
+
+        // see if there's another serial port configured for the correct protocol:
+        AP_HAL::UARTDriver *uart { AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_HobbyWing, num_escs) };
+        if (uart == nullptr) {
+            break;
+        }
+        switch (version) {
+        case 3:
+            escs[num_escs] = new AP_HobbyWing_Platinum_PRO_v3(*uart);
+            break;
+        case 4:
+            escs[num_escs] = new AP_HobbyWing_Platinum_v4(*uart);
+            break;
+        case 5:
+            escs[num_escs] = new AP_HobbyWing_XRotor_v4(*uart);
+            break;
+        }
+        if (escs[num_escs] == nullptr) {
+            break;
+        }
+        escs[num_escs]->set_poles(poles);
+        servo_channel[num_escs++] = i+1;
+    }
+
+    if (num_escs == 0) {
+        return;
+    }
+
+    // start a thread to handle the reading from all of those UARTs:
+    if (!hal.scheduler->thread_create(
+            FUNCTOR_BIND_MEMBER(&AP_ESC_Telem_HobbyWing::thread_main, void),
```

Comment:
I think this should be at the same priority as BLHeli - whatever that is

Reply:
Sounds right.  Perhaps this should share a thread with blheli?